expression/builtin_json.go

```diff
@@ -702,8 +703,115 @@ type jsonArrayAppendFunctionClass struct {
 	baseFunctionClass
 }
 
+type builtinJSONArrayAppendSig struct {
+	baseBuiltinFunc
+}
+
+func (c *jsonArrayAppendFunctionClass) verifyArgs(args []Expression) error {
+	if len(args) < 3 || (len(args)&1 != 1) {
+		return ErrIncorrectParameterCount.GenWithStackByArgs(c.funcName)
+	}
+	return nil
+}
+
 func (c *jsonArrayAppendFunctionClass) getFunction(ctx sessionctx.Context, args []Expression) (builtinFunc, error) {
-	return nil, errFunctionNotExists.GenWithStackByArgs("FUNCTION", "JSON_ARRAY_APPEND")
+	if err := c.verifyArgs(args); err != nil {
+		return nil, err
+	}
+	argTps := make([]types.EvalType, 0, len(args))
+	argTps = append(argTps, types.ETJson)
+	for i := 1; i < len(args)-1; i += 2 {
+		argTps = append(argTps, types.ETString, types.ETJson)
+	}
+	bf := newBaseBuiltinFuncWithTp(ctx, args, types.ETJson, argTps...)
+	for i := 2; i < len(args); i += 2 {
+		DisableParseJSONFlag4Expr(args[i])
+	}
+	sig := &builtinJSONArrayAppendSig{bf}
+	sig.setPbCode(tipb.ScalarFuncSig_JsonArrayAppendSig)
+	return sig, nil
+}
+
+func (b *builtinJSONArrayAppendSig) Clone() builtinFunc {
+	newSig := &builtinJSONArrayAppendSig{}
+	newSig.cloneFrom(&b.baseBuiltinFunc)
+	return newSig
+}
+
+func (b *builtinJSONArrayAppendSig) checkError(err error) error {
+	if err == nil || b.ctx.GetSessionVars().StrictSQLMode {
+		return err
+	}
+	b.ctx.GetSessionVars().StmtCtx.AppendWarning(err)
+	return nil
+}
+
+func (b *builtinJSONArrayAppendSig) evalJSON(row chunk.Row) (res json.BinaryJSON, isNull bool, err error) {
+	res, isNull, err = b.args[0].EvalJSON(b.ctx, row)
+	if err != nil {
+		return res, true, b.checkError(err)
+	}
+	if isNull {
+		return res, true, nil
+	}
+
+	for i := 1; i < len(b.args)-1; i += 2 {
+		// bad arguments should return null
+		// NOTE: mysql does not report error
+		// also they does not ignore it, they just return NULL
+		s, isNull, err := b.args[i].EvalString(b.ctx, row)
+		if isNull {
+			return res, true, b.checkError(json.ErrInvalidJSONPath.GenWithStackByArgs(s))
```

Comment:
This looks incorrect, if `isNull` is true, while `err` is nil, we should return `null` without error or warning?

Reply:
JSON path cannot be NULL, so it should return err

Reply:
It can be null when explicitly specified. In MySQL:
```
mysql> SELECT JSON_ARRAY_APPEND(@j, null, 1);
+--------------------------------+
| JSON_ARRAY_APPEND(@j, null, 1) |
+--------------------------------+
| NULL                           |
+--------------------------------+
```

Reply:
Well, I installed MariaDB before, and its behavior differs from MySQL. I will test them on MySQL 8.0.